keyboard shortcut

vsrisuknimit opened the pull request:
Here is a draft of the keyboard shortcuts.

It does not listen yet or not redirect. I cannot tell which is which. I look into event listener on the browser but I don't see anything when a press a key. 

Tag @bafolts for help.

This can close #1879 .

kberg commented on src/components/PlayerHome.ts:
```diff
@@ -49,6 +50,16 @@ export const PlayerHome = Vue.component('player-home', {
+    navigatePage: function(event: any) {
+      console.log('key pressed');
+      if (event.keyCode === KeyboardNavigation.COLONIES) {
+        console.log('f pressed');
```
remove.

kberg commented on src/KeyboardNavigation.ts:
```diff
@@ -0,0 +1,12 @@
+
+export enum KeyboardNavigation {
```
What a great selection of shortcuts.

kberg reviewed (commented):
**UM.** How did I manage to push directly to your repo?

Anyway, I managed to get keyboard shortcuts sort of workingg, and somehow they're now in this PR because of it.

vsrisuknimit commented:
Thank @kberg and @bafolts  . I changed the getElementByClass to getElementById instead. That is more specific and redirect us to where we want. Changed the behavior of the scrollIntoView to put the element in the middle of the page instead of the very far top which is obscured by the top bar.

Do we need both mounted and v-on? I thought one of the them should do. I tried to remove v-on but it seems we need it.

One major problem is that now there is this little blue border going over the whole element with v-on. Not sure how to remove it.

![image](https://user-images.githubusercontent.com/14239220/104305393-a2ebef00-549a-11eb-8bfa-8533d6a0b9af.png)

vsrisuknimit commented:
I was able to remove the blue outline but we hate in-line style so that perhaps should be somewhere else. I am not sure where since :class for that element is doing something weird.

chosta commented:
@vsrisuknimit you can use a function for the classes based on conditional logic.
Do a global search for `getClasses` to see how this could be done.

vsrisuknimit commented:
@Lynesth raised the issue of non-qwerty keyboards. This mapping does not make a lot of sense there.

I think the best path forward is to have mappable keys. I will need help with that.

Otherwise we can do shift+1, shift+2, but one key is much better.

kberg commented:
> @Lynesth raised the issue of non-qwerty keyboards. This mapping does not make a lot of sense there.
> 
> I think the best path forward is to have mappable keys. I will need help with that.
> 
> Otherwise we can do shift+1, shift+2, but one key is much better.

No, but 1,2,3,4,5 are excellent

vsrisuknimit commented:
@kberg just plain number will not work well. Some users may type in a number when paying for a cost.

![image](https://user-images.githubusercontent.com/14239220/104352491-bf0c8200-54d4-11eb-9aaa-c2e406c4601d.png)

vsrisuknimit commented:
Tested. Typing in the input field will trigger the keyboard shortcut. So we should avoid plain number.

bafolts reviewed (changes requested):
This looks right. I would use `KeyboardEvent` instead of `any`.

vsrisuknimit commented on src/components/PlayerHome.ts:
```diff
@@ -74,6 +75,45 @@ export const PlayerHome = Vue.component('player-home', {
   },
   mixins: [PlayerMixin],
   methods: {
+    navigatePage: function(event: any) {
```
@bafolts if I use KeyboardEvent, event.target.nodeName does not work. "Property 'nodeName' does not exist on type 'EventTarget'."

vsrisuknimit replied:
Fixed now. Cast event.target as Element.

vsrisuknimit commented:
Added key mapping graphics in the help page.

![image](https://user-images.githubusercontent.com/14239220/104487323-f34a7600-559a-11eb-8478-6ebc160df2f5.png)

vsrisuknimit commented:
These are the poll results from Discord.

![image](https://user-images.githubusercontent.com/14239220/104488279-22adb280-559c-11eb-8df7-17df5850ff4b.png)

![image](https://user-images.githubusercontent.com/14239220/104488298-28a39380-559c-11eb-9187-6ffed9f7b6c3.png)

I also added the ability to change the scrolling speed in preferences setting.

bafolts commented on src/KeyboardNavigation.ts:
```diff
@@ -0,0 +1,12 @@
+
+export enum KeyboardNavigation {
+  GAMEBOARD = 'Digit1',
+  TURMOILBOARD = '',
```
Many of these are missing. I can't imagine this works with these missing.

vsrisuknimit replied:
4 are what I decided would be useful in the end. I left the rest there as a place holder in the future if we want remappable keys. Should I remove them all?

bafolts replied:
Yes. If we decide to add them later we can add them then. I would also remove their corresponding ids from the template.

vsrisuknimit replied:
Removed the unused keys and ids.

bafolts commented on src/components/PlayerHome.ts:
```diff
@@ -163,7 +208,7 @@ export const PlayerHome = Vue.component('player-home', {
   },
   template: `
-        <div id="player-home" :class="player.turmoil ? 'with-turmoil': ''">
+        <div v-on:keydown="navigatePage" tabindex="0" id="player-home" :class="'shortkey-no-outline'+(player.turmoil ? ' with-turmoil': '')">
```
I think you can use `v-on:keydown.prevent` instead of manually calling to prevent default ourselves.

vsrisuknimit replied:
keydown.prevent causes an error.

![image](https://user-images.githubusercontent.com/14239220/104534762-e0f42a80-55e2-11eb-8a2a-555d9a1e2bb8.png)
